pkg/sql/lease.go

```diff
@@ -1192,14 +1233,27 @@ func nameMatchesTable(table *sqlbase.TableDescriptor, dbID sqlbase.ID, tableName
 // the timestamp. It returns the table descriptor and a expiration time.
 // A transaction using this descriptor must ensure that its
 // commit-timestamp < expiration-time. Care must be taken to not modify
-// the returned descriptor.
+// the returned descriptor. Renewal of a lease may begin in the
+// background. Renewal is done in order to prevent blocking on future
+// acquisitions.
 func (m *LeaseManager) AcquireByName(
 	ctx context.Context, timestamp hlc.Timestamp, dbID sqlbase.ID, tableName string,
 ) (*sqlbase.TableDescriptor, hlc.Timestamp, error) {
 	// Check if we have cached an ID for this name.
 	tableVersion := m.tableNames.get(dbID, tableName, timestamp)
 	if tableVersion != nil {
 		if !timestamp.Less(tableVersion.ModificationTime) {
+
+			// Atomically check and begin a renewal if one has not already
+			// been set.
+			durationUntilExpiration := time.Duration(tableVersion.expiration.WallTime - hlc.UnixNano())
+			durationUntilRenewal := durationUntilExpiration - m.LeaseStore.leaseRenewalTimeout
+			if t := m.findTableState(tableVersion.ID, false /* create */); t != nil &&
+				durationUntilRenewal < 0 && atomic.CompareAndSwapInt32(&t.renewalInProgress, 0, 1) {
+				// Start the renewal. When it finishes, it will reset t.renewalInProgress.
+				t.stopper.RunWorker(context.Background(), t.renewalFn(m, tableVersion))
+			}
```

Comment:
RunWorker() is used to start goroutines that run forever. How about modifying this code as:

```
if time.Duration(tableVersion.expiration.WallTime - timestamp.WallTime) < m.LeaseStore.leaseRenewalTimeout {
   if t := m.findTableState(tableVersion.ID, false /* create */); t != nil && atomic.CompareAndSwapInt32(&t.renewalInProgress, 0, 1) {
 				// Start the renewal. When it finishes, it will reset t.renewalInProgress.
 				if err := t.stopper.RunTaskAsync(context.Background(), t.renewalFn(m, tableVersion)); err != {
return &tableVersion.TableDescriptor, tableVersion.expiration, err
}
 	}
}
```